### PR TITLE
Resolve the correct workspace for alert resource

### PR DESCRIPTION
**Pull request by ilaizi**
This change achieves two objectives:
1. All links to log analytics will be created with the correct workspace for the resource in the alert.
2. After the alert is returned by detector, ARM will be queried with the resource on which the detector ran and not with the resource in the alert. Before this change, ARM call used to fail in Dogfood, because subscription id of the resource in the alert is different from subscription id of the workspace.

### CODE REVIEW

**Comment by bermanguy** [comment]
>                 alerts = await smartDetector.AnalyzeResourcesAsync(analysisRequest, this.tracer, cancellationToken);

are you synced with master? I don't see the Task.Run here

---
Refers to: src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs:101 in 7b746c8. [](commit_id = 7b746c8ba2c121c86c26e56f4db7827539dec94a, deletion_comment = False)

**Comment by liadtal** [comment]
Following our discussion, please move the flow to the QueryInfoProvider. The flow should be:
    If resource is AI
        AI
    If resource is workspace
        workspace
    If resource is subscription
        all workspaces
    else
        create log analytics client for the subscription
        run query and use correct workspaces
        use caching to avoid repeating calls

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs [hunk 1]
```diff
@@ -167,17 +168,41 @@ public bool IsSmartDetectorRunning
                         DataEndTime = currentTime
                     };
 
+                    Dictionary<ResourceIdentifier, string> resourceToWorkspaceMapping = new Dictionary<ResourceIdentifier, string>();
+                    if (newAlerts.Any(alert => alert.ResourceIdentifier.ResourceType != ResourceType.ApplicationInsights))
+                    {
+                        ITelemetryDataClient dataClient = await this.analysisServicesFactory.CreateLogAnalyticsTelemetryDataClientAsync(targetResourcesForDetector, cancellationToken);
+                        resourceToWorkspaceMapping = await LogAnalyticsHelperMethods.CreateResourceToWorkspaceMapping(dataClient, currentTime, this.tracer, cancellationToken);
```
Comment (bermanguy) [inline src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: >tracer [](start = 148, length = 6)

take the property - this.Tracer

**Comment by bermanguy** [comment]
>                     this.tracer.TraceInformation($"Start analysis, end of time range: {currentTime}");

Please fix also here to this.Tracer

---
Refers to: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs:153 in 7b746c8. [](commit_id = 7b746c8ba2c121c86c26e56f4db7827539dec94a, deletion_comment = False)

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs [hunk 2]
```diff
@@ -167,17 +168,41 @@ public bool IsSmartDetectorRunning
...
+                        {
+                            if (resourceToWorkspaceMapping.ContainsKey(newAlert.ResourceIdentifier))
+                            {
+                                queryRunInfo = new QueryRunInfo
+                                {
+                                    Type = TelemetryDbType.LogAnalytics,
+                                    ResourceIds = new List<string> { resourceToWorkspaceMapping[newAlert.ResourceIdentifier] }
+                                };
+                            }
+                        }
```
Comment (bermanguy) [inline src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: consider extracting to a separate  method so this method won't be so long

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs [hunk 3]
```diff
@@ -167,17 +168,41 @@ public bool IsSmartDetectorRunning
...
+                        {
+                            queryRunInfo = new QueryRunInfo
+                            {
+                                ResourceIds = new List<string> { newAlert.ResourceIdentifier.ToResourceId() },
+                                Type = TelemetryDbType.ApplicationInsights
+                            };
+                        }
+                        else
+                        {
+                            if (resourceToWorkspaceMapping.ContainsKey(newAlert.ResourceIdentifier))
```
Comment (bermanguy) [inline src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: >if [](start = 28, length = 2)

what if not?

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs [hunk 4]
```diff
@@ -167,17 +168,41 @@ public bool IsSmartDetectorRunning
...
                     };
 
+                    Dictionary<ResourceIdentifier, string> resourceToWorkspaceMapping = new Dictionary<ResourceIdentifier, string>();
+                    if (newAlerts.Any(alert => alert.ResourceIdentifier.ResourceType != ResourceType.ApplicationInsights))
+                    {
+                        ITelemetryDataClient dataClient = await this.analysisServicesFactory.CreateLogAnalyticsTelemetryDataClientAsync(targetResourcesForDetector, cancellationToken);
+                        resourceToWorkspaceMapping = await LogAnalyticsHelperMethods.CreateResourceToWorkspaceMapping(dataClient, currentTime, this.tracer, cancellationToken);
+                    }
+
                     // Show the alerts that were found in this iteration
```
Comment (bermanguy) [inline src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: >// Show the alerts that were found in this iteration [](start = 20, length = 52)

I think you can remove this comment, it's not really helping (my bad)

File: src/SDK/SmartDetectorsShared/Clients/AzureResourceManagerClient.cs [hunk 5]
```diff
@@ -235,7 +235,7 @@ public Task<ResourceProperties> GetResourcePropertiesAsync(ResourceIdentifier re
         {
             if (resourceIdentifier.ResourceType != ResourceType.LogAnalytics)
             {
-                throw new ArgumentOutOfRangeException(nameof(resourceIdentifier.ResourceType), resourceIdentifier.ResourceType, "The resource type must be ApplicationInsights");
+                throw new ArgumentOutOfRangeException(nameof(resourceIdentifier.ResourceType), resourceIdentifier.ResourceType, "The resource type must be LogAnalytics");
```
Comment (bermanguy) [inline src/SDK/SmartDetectorsShared/Clients/AzureResourceManagerClient.cs]: >LogAnalytics [](start = 155, length = 12)

maybe replace with something like - ResourceType.LogAnalytics.ToString()

File: src/SDK/SmartDetectorsShared/Clients/LogAnalyticsHelperMethods.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        /// This method is useful in cases when the data client can access multiple workspaces
+        /// </summary>
+        /// <param name="dataClient">The Log Analytics client for querying the data</param>
+        /// <param name="dataEndTime">The data end time</param>
+        /// <param name="tracer">The tracer</param>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns>A task returning a mapping between resources and workspaces that these resources are connected to</returns>
+        public static async Task<Dictionary<ResourceIdentifier, string>> CreateResourceToWorkspaceMapping(ITelemetryDataClient dataClient, DateTime dataEndTime, ITracer tracer, CancellationToken cancellationToken)
+        {
+            if (dataClient is LogAnalyticsTelemetryDataClient == false)
```
Comment (bermanguy) [inline src/SDK/SmartDetectorsShared/Clients/LogAnalyticsHelperMethods.cs]: >dataClient is LogAnalyticsTelemetryDataClient [](start = 16, length = 45)

maybe (!dataClient is LogAnalyticsTelemetryDataClient) more readable?

File: src/SDK/SmartDetectorsShared/Clients/LogAnalyticsHelperMethods.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        /// Creates a mapping between resources and workspaces that these resources are connected to.
+        /// The workspaces must be accessible by the <see cref="dataClient"/> in order for them (and resources connected to them) to appear in the mapping.
+        /// This method is useful in cases when the data client can access multiple workspaces
+        /// </summary>
+        /// <param name="dataClient">The Log Analytics client for querying the data</param>
+        /// <param name="dataEndTime">The data end time</param>
+        /// <param name="tracer">The tracer</param>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns>A task returning a mapping between resources and workspaces that these resources are connected to</returns>
+        public static async Task<Dictionary<ResourceIdentifier, string>> CreateResourceToWorkspaceMapping(ITelemetryDataClient dataClient, DateTime dataEndTime, ITracer tracer, CancellationToken cancellationToken)
```
Comment (bermanguy) [inline src/SDK/SmartDetectorsShared/Clients/LogAnalyticsHelperMethods.cs]: >CreateResourceToWorkspaceMapping [](start = 73, length = 32)

UTs are necessary?

File: src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs [hunk 8]
```diff
@@ -123,11 +123,38 @@ public class SmartDetectorRunner : ISmartDetectorRunner
                 this.tracer.ReportMetric("AlertType", alertType.Count(), new Dictionary<string, string>() { { "AlertType", alertType.Key } });
             }
 
+            Dictionary<ResourceIdentifier, string> resourceToWorkspaceMapping = new Dictionary<ResourceIdentifier, string>();
```
Comment (EranGon) [inline src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs]: >Dictionary [](start = 12, length = 10)

add a comment here

File: src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs [hunk 9]
```diff
@@ -123,11 +123,38 @@ public class SmartDetectorRunner : ISmartDetectorRunner
...
+            if (alerts.Any(alert => alert.ResourceIdentifier.ResourceType != ResourceType.ApplicationInsights))
+            {
+                ITelemetryDataClient dataClient = await this.analysisServicesFactory.CreateLogAnalyticsTelemetryDataClientAsync(resources, cancellationToken);
+                resourceToWorkspaceMapping = await LogAnalyticsHelperMethods.CreateResourceToWorkspaceMapping(dataClient, request.DataEndTime, this.tracer, cancellationToken);
```
Comment (EranGon) [inline src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs]: >LogAnalyticsHelperMethods [](start = 51, length = 25)

why do you need this class? it's only used here, so put this as a private method in the runner class

File: src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs [hunk 10]
```diff
@@ -123,11 +123,38 @@ public class SmartDetectorRunner : ISmartDetectorRunner
...
+                {
+                    if (resourceToWorkspaceMapping.ContainsKey(alert.ResourceIdentifier))
+                    {
+                        queryRunInfo = new QueryRunInfo
+                        {
+                            Type = TelemetryDbType.LogAnalytics,
+                            ResourceIds = new List<string> { resourceToWorkspaceMapping[alert.ResourceIdentifier] }
+                        };
+                    }
+                }
```
Comment (EranGon) [inline src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs]: why not encapsulate all of this code inside queryRunInfoProvider.GetQueryRunInfoAsync?

File: src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs [hunk 11]
```diff
@@ -123,11 +123,38 @@ public class SmartDetectorRunner : ISmartDetectorRunner
...
+                {
+                    queryRunInfo = new QueryRunInfo
+                    {
+                        ResourceIds = new List<string> { alert.ResourceIdentifier.ToResourceId() },
+                        Type = TelemetryDbType.ApplicationInsights
+                    };
+                }
+                else
+                {
+                    if (resourceToWorkspaceMapping.ContainsKey(alert.ResourceIdentifier))
```
Comment (EranGon) [inline src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs]: >if [](start = 20, length = 2)

no else?

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs [hunk 12]
```diff
@@ -167,17 +168,41 @@ public bool IsSmartDetectorRunning
...
+                        QueryRunInfo queryRunInfo = null;
+                        if (newAlert.ResourceIdentifier.ResourceType == SmartDetectors.ResourceType.ApplicationInsights)
+                        {
+                            queryRunInfo = new QueryRunInfo
+                            {
+                                ResourceIds = new List<string> { newAlert.ResourceIdentifier.ToResourceId() },
+                                Type = TelemetryDbType.ApplicationInsights
+                            };
+                        }
+                        else
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: isn't this a lot of code duplication from the regular runner?

**Comment by EranGon** [comment]
>                     this.tracer.TraceInformation($"Start analysis, end of time range: {currentTime}");

or better yet - remove the lower case tracer member, and make Tracer a get only property

---
In reply to: [388627659](https://github.com/liadtal/Smart-Alerts/pull/152#issuecomment-388627659) [](ancestors = 388627659)

---
Refers to: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs:153 in 7b746c8. [](commit_id = 7b746c8ba2c121c86c26e56f4db7827539dec94a, deletion_comment = False)

File: src/SDK/SmartDetectorsShared/Clients/AzureResourceManagerClient.cs (hunk 5 above)
Comment (EranGon) [reply src/SDK/SmartDetectorsShared/Clients/AzureResourceManagerClient.cs]: or $"The resource type must be {ResourceType.LogAnalytics}"

---
In reply to: [187801610](https://github.com/liadtal/Smart-Alerts/pull/152#discussion_r187801610) [](ancestors = 187801610)

File: src/SDK/SmartDetectorsShared/Clients/LogAnalyticsHelperMethods.cs (hunk 6 above)
Comment (EranGon) [reply src/SDK/SmartDetectorsShared/Clients/LogAnalyticsHelperMethods.cs]: both are a violation of the abstraction that ITelemetryDataClient provides

a better way would be to add the TelemetryDbType as a property to the interface

---
In reply to: [187801705](https://github.com/liadtal/Smart-Alerts/pull/152#discussion_r187801705) [](ancestors = 187801705)

File: src/SDK/SmartDetectorsShared/Clients/LogAnalyticsHelperMethods.cs (hunk 7 above)
Comment (EranGon) [reply src/SDK/SmartDetectorsShared/Clients/LogAnalyticsHelperMethods.cs]: the most certainly are needed

---
In reply to: [187801883](https://github.com/liadtal/Smart-Alerts/pull/152#discussion_r187801883) [](ancestors = 187801883)

File: src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs (hunk 9 above)
Comment (ilaizi) [reply src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs]: It's also used in emulator SmartDetectorRunner

---
In reply to: [187858588](https://github.com/liadtal/Smart-Alerts/pull/152#discussion_r187858588) [](ancestors = 187858588)

File: src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs (hunk 10 above)
Comment (ilaizi) [reply src/analysis/SmartDetectorsAnalysis/SmartDetectorRunner.cs]: Discussed offline

---
In reply to: [187858765](https://github.com/liadtal/Smart-Alerts/pull/152#discussion_r187858765) [](ancestors = 187858765)

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs [hunk 13]
```diff
@@ -89,21 +93,9 @@ public ObservableCollection<EmulationAlert> Alerts
...
-                return this.tracer;
-            }
-
-            set
-            {
-                this.tracer = value;
-                this.OnPropertyChanged();
-            }
-        }
+        public ITracer Tracer { get; }
```
Comment (liadtal) [inline src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: >public [](start = 8, length = 6)

Why does it need to be public property?

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs [hunk 14]
```diff
@@ -256,5 +246,123 @@ private List<ResourceIdentifier> GetTargetResourcesForDetector(HierarchicalResou
...
+            {
+                return new QueryRunInfo
+                {
+                    ResourceIds = new List<string> { alert.ResourceIdentifier.ToResourceId() },
+                    Type = TelemetryDbType.ApplicationInsights
+                };
+            }
+            else
+            {
+                var resourceToWorkspaceResourceIdMapping = await lazyResourceToWorkspaceResourceIdMapping.Value;
```
Comment (liadtal) [inline src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: >var resourceToWorkspaceResourceIdMapping = await lazyResourceToWorkspaceResourceIdMapping.Value; [](start = 16, length = 96)

Please add comment explaining this logic

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs [hunk 15]
```diff
@@ -256,5 +246,123 @@ private List<ResourceIdentifier> GetTargetResourcesForDetector(HierarchicalResou
...
+        private async Task<QueryRunInfo> CreateQueryRunInfoForAlertAsync(SmartDetectors.Alert alert, Lazy<Task<Dictionary<ResourceIdentifier, ResourceIdentifier>>> lazyResourceToWorkspaceResourceIdMapping)
+        {
+            if (alert.ResourceIdentifier.ResourceType == SmartDetectors.ResourceType.ApplicationInsights)
+            {
+                return new QueryRunInfo
+                {
+                    ResourceIds = new List<string> { alert.ResourceIdentifier.ToResourceId() },
+                    Type = TelemetryDbType.ApplicationInsights
+                };
+            }
```
Comment (liadtal) [inline src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: This case can call the queryInfoProvider (e.g, reach the "else" below)

File: test/SDK/MonitoringApplianceEmulatorTests/SmartDetectorRunnerTests.cs [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Monitoring.SmartDetectors;
+    using Microsoft.Azure.Monitoring.SmartDetectors.Clients;
+    using Microsoft.Azure.Monitoring.SmartDetectors.MonitoringAppliancEmulator.Models;
+    using Microsoft.Azure.Monitoring.SmartDetectors.State;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+    using Moq;
+
+    [TestClass]
+    public class SmartDetectorRunnerTests
```
Comment (bermanguy) [inline test/SDK/MonitoringApplianceEmulatorTests/SmartDetectorRunnerTests.cs]: >SmartDetectorRunnerTests [](start = 17, length = 24)

Thanks for adding it!!! You are definitely an Emulator owner now :)

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs [hunk 17]
```diff
@@ -167,24 +160,21 @@ public bool IsSmartDetectorRunning
                         DataEndTime = currentTime
                     };
 
-                    // Show the alerts that were found in this iteration
-                    newAlerts.ForEach(async newAlert =>
+                    var lazyResourceToWorkspaceResourceIdMapping = new Lazy<Task<Dictionary<ResourceIdentifier, ResourceIdentifier>>>(() => this.GetResourceToWorkspaceResourceIdMappingAsync(targetResourcesForDetector, currentTime, this.Tracer, cancellationToken));
```
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: >Lazy<Task [](start = 71, length = 9)

use AsyncLazy

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs [hunk 18]
```diff
@@ -256,5 +246,123 @@ private List<ResourceIdentifier> GetTargetResourcesForDetector(HierarchicalResou
...
             return targetResourcesForDetector;
         }
+
+        /// <summary>
+        /// Creates <see cref="QueryRunInfo"/> for alert based on alert resource id and a mapping between resources and workspaces
+        /// </summary>
+        /// <param name="alert">The alert</param>
+        /// <param name="lazyResourceToWorkspaceResourceIdMapping">Lazily computed mapping between resources and workspaces</param>
+        /// <returns>A task returning <see cref="QueryRunInfo"/> for alert</returns>
+        private async Task<QueryRunInfo> CreateQueryRunInfoForAlertAsync(SmartDetectors.Alert alert, Lazy<Task<Dictionary<ResourceIdentifier, ResourceIdentifier>>> lazyResourceToWorkspaceResourceIdMapping)
```
[the remark this answers is not in the document]
Comment (EranGon) [reply src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: why? do you expect it to be reused?

---
In reply to: [188943303](https://github.com/liadtal/Smart-Alerts/pull/152#discussion_r188943303) [](ancestors = 188943303)

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs (hunk 14 above)
Comment (EranGon) [inline src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: >var [](start = 16, length = 3)

don't use var when the type is not explicit from the line of code please

File: src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs (hunk 17 above)
Comment (ilaizi) [reply src/SDK/MonitoringApplianceEmulator/Models/SmartDetectorRunner.cs]: I don't believe AsyncLazy exists in .NET

---
In reply to: [188956203](https://github.com/liadtal/Smart-Alerts/pull/152#discussion_r188956203) [](ancestors = 188956203)

**Pull request merged**